crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4640,10 +4650,413 @@ mod tests {
+            else:
+                x = 13
+                y = 23
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // TODO: What would be great in these assertions would be
+        // if we could assert the types of values *within* the `else` block.
+        // (Not possible with our current testing infrastructure, sadly.) --Alex
+
+        assert_public_ty(&db, "src/a.py", "x", "Literal[11, 12, 13]");
+
+        // `y` may be unbound if the `except NameError` branch is taken:
+        assert_public_ty(&db, "src/a.py", "y", "Unbound | Literal[2, 22, 23]");
+
+        // `z` may be unbound if the `except NameError` or the `except TypeError` branch is taken:
+        assert_public_ty(&db, "src/a.py", "z", "Unbound | Literal[3]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn nested_try_except_definitions() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                try:
+                    try:
+                        x = 1
+                    except TypeError:
+                        x = 2
+                    else:
+                        x = 3
+                except RuntimeError:
+                    x = 4
+                else:
+                    x = 5
+            except TypeError:
+                try:
+                    try:
+                        x = 6
+                    except RuntimeError:
+                        x = 7
+                    else:
+                        x = 8
+                except OSError:
+                    x = 9
+                else:
+                    x = 10
+            else:
+                try:
+                    try:
+                        x = 11
+                    except RuntimeError:
+                        x = 12
+                    else:
+                        x = 13
+                except OSError:
+                    x = 14
+                else:
+                    x = 15
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+        assert_public_ty(
+            &db,
+            "src/a.py",
+            "x",
+            "Literal[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]",
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_multiple_definitions_in_exhaustive_nested_try_block() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                try:
+                    x = 1
+                except:
+                    x = 2
+                y = 11
+            except TypeError:
+                x = 101
+                y = 111
+            except:
+                y = 211
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+        assert_public_ty(&db, "src/a.py", "x", "Unbound | Literal[1, 2, 101]");
```

Comment:
I don't see how `x` could actually be unbound in this scenario, but I'm also not sure if it's worth fixing. We assume that we could exit the outer try with an exception before executing the inner try at all, but that's not actually true in this example; nothing happens in the outer try before we enter the inner try, so in actual fact we are guaranteed to either complete the assignment `x = 1` in the inner try, or have an exception caught by the exhaustive inner handler, which also binds `x`.

I think we would handle this correctly if, instead of saving a pre-try state, and then pushing a state post each definition, we instead pushed a state pre each definition, and then saved a post-try state.

Reply:
I think this comment is wrong; given that we don't currently try to distinguish non-raising statements, we have to assume that `x = 2` in the inner exhaustive `except` could also raise, so it is correct for `x` to be possibly unbound.